server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -50,187 +52,267 @@
+    if (params.getAvailableSegments().contains(segment)) {
+      assign(targetReplicants, currentReplicants, params, segment, stats);
+    }
 
-    final Map<String, Integer> loadStatus = Maps.newHashMap();
+    drop(targetReplicants, currentReplicants, params, segment, stats);
 
-    int totalReplicantsInCluster = params.getSegmentReplicantLookup().getTotalReplicants(segment.getIdentifier());
-    for (Map.Entry<String, Integer> entry : getTieredReplicants().entrySet()) {
-      final String tier = entry.getKey();
-      final int expectedReplicantsInTier = entry.getValue();
-      final int totalReplicantsInTier = params.getSegmentReplicantLookup()
-                                              .getTotalReplicants(segment.getIdentifier(), tier);
-      final int loadedReplicantsInTier = params.getSegmentReplicantLookup()
-                                               .getLoadedReplicants(segment.getIdentifier(), tier);
+    return stats;
+  }
+
+  private static void assign(
+      final Object2IntMap<String> targetReplicants,
+      final Object2IntMap<String> currentReplicants,
+      final DruidCoordinatorRuntimeParams params,
+      final DataSegment segment,
+      final CoordinatorStats stats
+  )
+  {
+    // if primary replica already exists
+    if (!currentReplicants.isEmpty()) {
+      assignReplicas(targetReplicants, currentReplicants, params, segment, stats);
+    } else {
+      final ServerHolder primaryHolderToLoad = assignPrimary(targetReplicants, params, segment);
+      if (primaryHolderToLoad == null) {
+        // cluster does not have any replicants and cannot identify primary holder
+        // this implies that no assignment could be done
+        return;
+      }
+
+      final String tier = primaryHolderToLoad.getServer().getTier();
+      // assign replicas for the rest of the tier
+      final int numAssigned = 1 /* primary */ + assignReplicasForTier(
+          tier,
+          targetReplicants.getOrDefault(tier, 0),
+          currentReplicants.getOrDefault(tier, 0) + 1,
+          params,
+          getHolderList(
+              tier,
+              params.getDruidCluster(),
+              createPredicate(params),
+              holder -> !holder.equals(primaryHolderToLoad)
+          ),
+          segment
+      );
+      stats.addToTieredStat(ASSIGNED_COUNT, tier, numAssigned);
+
+      // tier with primary replica
+      final int targetReplicantsInTier = targetReplicants.removeInt(tier);
+      // assign replicas for the other tiers
+      assignReplicas(targetReplicants, currentReplicants, params, segment, stats);
+      // add back tier (this is for processing drop)
+      targetReplicants.put(tier, targetReplicantsInTier);
+    }
+  }
 
-      final MinMaxPriorityQueue<ServerHolder> serverQueue = params.getDruidCluster().getHistoricalsByTier(tier);
+  private static Predicate<ServerHolder> createPredicate(final DruidCoordinatorRuntimeParams params)
```

Comment:
Could this method be given a more speaking name? E. g. "createLoadQueueSizeLimitingPredicate()"

Reply:
👍